### PR TITLE
refactor(pie-storybook): DSW-1218 make story creation more generic

### PR DESCRIPTION
**Pull request by xander-marjoram**
## Describe your changes (can list changeset entries if preferable)
- Modal and cookie banner are still to-do, but I'll attempt this in a following PR. They're a bit more complicated because they have multiple templates.

## Author Checklist (complete before requesting a review)
- [x] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests
- [ ] If it is a `PIE Docs` change, I have reviewed the Docs site preview
- [x] If it is a component change, I have reviewed the Storybook preview
- [ ] If there are visual test updates, I have reviewed them properly before approving

### Reviewer 1
- [ ] If it is a `PIE Docs` change, I have reviewed the PR preview
- [ ] If there are visual test updates, I have reviewed them

### Reviewer 2
- [ ] If it is a `PIE Docs` change, I have reviewed the PR preview
- [ ] If there are visual test updates, I have reviewed them

### CODE REVIEW

File: apps/pie-storybook/types/StoryOptions.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+type BackgroundColor = 'light (container-default)' |
+    'dark (container-dark)' |
+    'background-subtle' |
+    'brand orange' |
+    'saddlebrown' |
```
Comment (jamieomaguire) [inline apps/pie-storybook/types/StoryOptions.ts:5]: nice

Comment (xander-marjoram) [reply apps/pie-storybook/types/StoryOptions.ts:5]: I was so preoccupied with whether I could, I didn't stop to think if I should.

File: apps/pie-storybook/types/StoryOptions.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    'aquamarine';
+
+type BackgroundValue = {
+    name: BackgroundColor,
+    value: string,
```
Comment (jamieomaguire) [inline apps/pie-storybook/types/StoryOptions.ts:10]: is it worth union typing the values?

File: apps/pie-storybook/utilities/index.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { TemplateResult } from 'lit';
+import { StoryOptions } from '../types/StoryOptions';
+
+export type TemplateFunction<T> = (props: T) => TemplateResult;
+
+export const createStory = <T>(templateFunc: TemplateFunction<T>, defaultArgs: T) => (
```
Comment (jamieomaguire) [inline apps/pie-storybook/utilities/index.ts:6]: I think JSDocs would go a long way here

File: apps/pie-storybook/types/StoryOptions.ts (hunk 2 above)
Comment (xander-marjoram) [reply apps/pie-storybook/types/StoryOptions.ts:10]: I did briefly consider this, but I'm pretty sure they're only used in that one object so I'm not sure there's much value in it, what do you think?
I did just notice one of the imports is wrong in another file so lemme fix that.

Comment (jamieomaguire) [reply apps/pie-storybook/types/StoryOptions.ts:10]: It seems a bit odd that we could add a key of `saddlebrown` with a value of `container-default` if that makes sense?

**Review by jamieomaguire — DISMISSED** [review]

File: apps/pie-storybook/types/StoryOptions.ts (hunk 2 above)
Comment (xander-marjoram) [reply apps/pie-storybook/types/StoryOptions.ts:10]: We paired on this and couldn't quite find a nice improvement, so we'll leave this for now.

File: apps/pie-storybook/stories/pie-toggle-switch.stories.ts [hunk 4]
```diff
@@ -95,11 +95,10 @@ const Template = (props: ToggleSwitchProps): TemplateResult => {
...
-            ?isDisabled="${isDisabled}" 
+            ?isDisabled="${isDisabled}"
         />`;
 };
 
-export const Default: Story<ToggleSwitchProps> = (args: ToggleSwitchProps) => Template(args);
-Default.args = {
-    ...defaultArgs,
-};
+const createToggleSwitchStory = createStory(Template, defaultArgs);
```
[the remark this answers is not in the document]
Comment (xander-marjoram) [reply apps/pie-storybook/stories/pie-toggle-switch.stories.ts:102]: Chatted with Jamie and he's happy for this to be resolved. If/when we start adding more stories this'll be useful.

**Review by jamieomaguire — APPROVED** [review]

File: apps/pie-storybook/types/StoryOptions.ts (hunk 1 above)
Comment (ashleynolan) [reply apps/pie-storybook/types/StoryOptions.ts:5]: I wouldn't approve the PR without this tbh

**Review by ashleynolan — APPROVED** [review]

**Pull request merged**